File: src/test/java/br/com/blz/testjava/api/product/ProductServiceTest.java
package br.com.blz.testjava.api.product;

import br.com.blz.testjava.api.mapper.MapperUtil;
import br.com.blz.testjava.api.product.controller.domain.ProductRequest;
import br.com.blz.testjava.api.product.controller.domain.ProductResponse;
import br.com.blz.testjava.api.product.entity.ProductEntity;
import br.com.blz.testjava.api.product.repository.ProductRepository;
import br.com.blz.testjava.api.product.service.ProductService;
import br.com.blz.testjava.common.exceptions.ProductAlreadyExistsException;
import br.com.blz.testjava.common.exceptions.ProductNotFoundException;
import br.com.blz.testjava.testUtils.TestUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ProductServiceTest {
  @Mock private ProductRepository repository;
  @Mock private MapperUtil mapper;
  @InjectMocks private ProductService productService;

  @Test
  void create_product_should_succeed() throws ProductAlreadyExistsException {
    ProductRequest productRequest = TestUtils.getProductRequest();
    ProductEntity productEntity = new ProductEntity();
    ProductResponse productResponse = TestUtils.getProductResponse();
    when(repository.save(productEntity)).thenReturn(productEntity);
    when(mapper.productRequestToEntity(productRequest)).thenReturn(productEntity);
    when(mapper.productEntitytoResponse(productEntity)).thenReturn(productResponse);

    ProductResponse response = productService.create(productRequest);

    assertEquals(response, productResponse);
  }

  @Test
  void get_product_should_succeed() throws ProductNotFoundException {
    ProductEntity productEntity = TestUtils.getProductEntity();
    ProductResponse productResponse = TestUtils.getProductResponse();
    when(repository.findBySku(1)).thenReturn(Optional.of(productEntity));
    when(mapper.productEntitytoResponse(productEntity)).thenReturn(productResponse);

    ProductResponse response = productService.getProduct(1);

    assertEquals(response, productResponse);
  }

  @Test
  void getProduct_whenProductExists_shouldReturnProduct() throws ProductNotFoundException {
    // Arrange
    ProductEntity productEntity = new ProductEntity();
    ProductResponse expectedProductResponse = new ProductResponse();
    when(repository.findBySku(anyInt())).thenReturn(Optional.of(productEntity));
    when(mapper.productEntitytoResponse(productEntity)).thenReturn(expectedProductResponse);

    // Act
    ProductResponse actualProductResponse = productService.getProduct(123);

    // Assert
    assertEquals(expectedProductResponse, actualProductResponse);
  }

  @Test
  void getProduct_whenProductDoesNotExist_shouldThrowProductNotFoundException() {
    // Arrange
    when(repository.findBySku(anyInt())).thenReturn(Optional.empty());
    // Act & Assert
    assertThrows(ProductNotFoundException.class, () -> productService.getProduct(123));
  }
}

File: src/test/java/br/com/blz/testjava/product/service/ProductServiceTest.kt
package br.com.blz.testjava.product.service

import br.com.blz.testjava.exception.DomainBusinessException
import br.com.blz.testjava.exception.EntityNotFoundException
import br.com.blz.testjava.product.mapper.ProductMapper
import br.com.blz.testjava.product.repository.IProductRepository
import br.com.blz.testjava.product.service.provider.ProductProvider
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import java.util.Optional.empty
import java.util.Optional.of

@ExtendWith(MockitoExtension::class)
internal class ProductServiceTest {

  @InjectMocks
  lateinit var productService: ProductService

  @Mock
  lateinit var productRepository: IProductRepository

  @Test
  fun `should throw error when trying to save existing product`() {
    val productInDTO = ProductProvider.createInDto()

    `when`(productRepository.existsBySku(productInDTO.sku)).thenReturn(true)

    assertThatExceptionOfType(DomainBusinessException::class.java).isThrownBy {
      productService.create(productInDTO)
    }.withMessageContaining("Não foi possível criar, produto já existente")
  }

  @Test
  fun `must create product`() {
    val productInDTO = ProductProvider.createInDto()

    `when`(productRepository.existsBySku(productInDTO.sku)).thenReturn(false)

    productService.create(productInDTO)

    verify(productRepository, times(1)).save(any())
  }

  @Test
  fun `must edit product by sku`() {
    val productInDTO = ProductProvider.createInDto()

    productService.editBySku(productInDTO.sku, productInDTO)

    verify(productRepository, times(1)).save(any())
  }

  @Test
  fun `must find product by sku`() {
    val product = ProductProvider.createEntity()

    `when`(productRepository.findBySku(product.sku)).thenReturn(of(product))

    val productOutDTO = productService.findBySku(product.sku)

    assertEquals(ProductMapper.toOutDto(product), productOutDTO)
  }

  @Test
  fun `should throw error when not finding product by sku`() {
    val sku = 43264L

    `when`(productRepository.findBySku(sku)).thenReturn(empty())

    assertThatExceptionOfType(EntityNotFoundException::class.java).isThrownBy {
      productService.findBySku(sku)
    }.withMessageContaining("Produto não encontrado")
  }

  @Test
  fun `must delete product by sku`() {
    val sku = 43264L

    productService.deleteBySku(sku)

    verify(productRepository, times(1)).deleteBySku(sku)
  }
}
